src/odemis/dataio/tiff.py

```diff
@@ -2722,3 +2742,192 @@ def _convert_thermo_fisher_to_odemis_metadata(metadata: str) -> dict:
         logging.warning(f"Failed to parse ThermoFisher metadata: {model.MD_DET_TYPE} - {e}")
 
     return md
+
+### Tescan Metadata parsing
+def _parse_tescan_metadata(metadata: bytes) -> str:
+    """
+    Parses TESCAN metadata bytes to str
+    :param metadata: metadata bytes (TESCAN bytes format)
+    :return: (str) metadata (TESCAN ini format)
+    """
+    # NOTE:
+    # METADATA FORMAT
+    # JPEG2000 metadata
+    #   4 bytes: length of the next chunk (n)
+    #   2 bytes: header of the next chunk (type)
+    #   n bytes: chunk data
+    # MAIN metadata
+    #   4 bytes: length of the next chunk (n)
+    #   2 bytes: header of the next chunk (type)
+    #   n - 3 bytes: valid chunk data
+    #   3 bytes: unsure?
+    # SEM/FIB metadata
+    #   4 bytes: length of the next chunk (n)
+    #   2 bytes: header of the next chunk (type)
+    #   n - 3 bytes: valid chunk data
+    #   3 bytes: unsure?
+
+    # constants
+    LENGTH_BYTES = 4
+    HEADER_BYTES = 2
+    END_BYTES = 3
+
+    # valid metadata str
+    vmd = ""
+
+    # header type map
+    header_map ={
+        2: "[MAIN]",
+        3: "[SEM]",
+        5: "[FIB]"
+    }
+
+    # loop through the chunks, extract the valid data
+    while len(metadata) > LENGTH_BYTES + HEADER_BYTES:
+        
+        # read the length of the next chunk (first 4 bytes)
+        chunk_len = int.from_bytes(metadata[:LENGTH_BYTES], 
+                                            byteorder="little", signed=False)
+        
+        # read the type of the chunk (next 2 bytes)
+        chunk_type = int.from_bytes(metadata[LENGTH_BYTES:LENGTH_BYTES + HEADER_BYTES], 
+                                            byteorder="little", signed=False)
+        if chunk_type == 0:
+            break # end of metadata
+
+        # read the chunk data as string (next n-3 bytes)
+        chunk_start = LENGTH_BYTES + HEADER_BYTES
+        chunk_end = chunk_start + chunk_len - END_BYTES
+        chunk = metadata[chunk_start:chunk_end].decode("latin1")
+
+        # read the final 3 bytes (unused)
+        # NOTE: 3 bytes at the end of each chunk, not sure what for?     
+        chunk_final = metadata[chunk_end:chunk_end + END_BYTES]
+
+        # add the chunk data to the metadata string (if valid)
+        if chunk_type in header_map:
+            vmd += f"{header_map[chunk_type]}\n" + chunk
+            print(chunk_type, chunk_final, metadata[chunk_end - 10:chunk_start + chunk_len ])
+        
+        # skip to next chunk
+        metadata = metadata[LENGTH_BYTES + chunk_len:] # skip to next chunk         
```

Comment:
I think there is a couple of errors with the indices. This code works fine, and there is not "end bytes" trick needed:
```suggestion
        # read the chunk data as string
        chunk_start = LENGTH_BYTES + HEADER_BYTES
        chunk_end = LENGTH_BYTES + chunk_len
        chunk = metadata[chunk_start:chunk_end]

        # add the chunk data to the metadata string (if valid), not including the null byte at the end of the string
        if chunk_type in header_map:
            vmd += header_map[chunk_type] + "\n" + chunk[:-1].decode("latin1")

        # skip to next chunk
        metadata = metadata[chunk_end:]  # skip to next chunk
```

Reply:
much cleaner, thanks 👍